### PR TITLE
[Perf][MOSS-TTS] Refactor and optimize vocoder decoding

### PR DESCRIPTION
**Pull request by CloudRipple**
## Motivation

`BatchVocoderBase` provides batched scheduling, but MOSS-TTS Delay still decodes every request and every audio segment independently. Each segment repeats inverse-delay materialization, quantizer work, decoder launches, and waveform transfer, leaving small codec operations and host synchronization on the critical path.

MOSS-TTS Local already has a packed non-streaming decoder, but that implementation is scoped to the Local package even though the legacy and v2 MOSS audio tokenizers expose equivalent decoder primitives. This PR moves the decoder to the shared MOSS package and extends it to support both codec layouts, allowing Delay to batch mixed-length segments and use packed FlashAttention without duplicating the Local implementation.

The goal is to accelerate non-streaming vocoder decoding while preserving FP32 quantizer semantics, waveform output, request ordering, streaming behavior, and a standalone codec fallback for unsupported configurations.

## Modifications

- Move the packed vocoder decoder from `moss_tts_local` to the shared `moss_tts` package and use neutral `MossAudioTokenizer*` names for both the legacy MOSS audio tokenizer and MOSS-Audio-Tokenizer-v2.
- Add cross-request segment batching for MOSS-TTS Delay:
  - extract and flatten every request's valid audio-code segments while retaining their request/segment order;
  - pad mixed-length codes once on CPU and issue one codes transfer per segment batch;
  - derive decoder output lengths on the host and copy only valid waveform samples back to CPU;
  - reassemble decoded segments into their original requests.
- Keep source-compatible codec numerics:
  - retain FP32 codebooks, projections, accumulation order, and waveform outputs in the quantizer path;
  - run only supported decoder stages under configurable lower-precision autocast;
  - use an exact interleaved RoPE kernel before packed FlashAttention;
  - preserve local-causal attention semantics by packing FA3 query tiles with the corresponding key windows.
- Remove avoidable hot-path work:
  - reverse the delay layout with a validated strided view instead of per-codebook slice copies;
  - cache inference-only quantizer tensors;
  - avoid device scalar reads when packing and unpacking mixed-length decoder inputs;
  - compact valid waveform ranges before the device-to-host transfer.
- Gate the optimized path on codec structure, device, dtype, and FlashAttention support. The custom RoPE kernel has its own layout/capability guard and exact fallback; initialization incompatibilities keep the standalone codec path, and a runtime batch failure disables the optimized path before retrying requests through the source decoder.
- Keep incremental streaming on the existing standalone codec route. MOSS-TTS Local imports the shared decoder directly only for its non-streaming packed path.
- Split non-streaming vocoder, decoder, quantizer, delay-pattern, fallback, ordering, dtype, and numerical-parity tests into focused MOSS-TTS test modules.

## Related Issues

#1233 

## Accuracy Test

This PR does not change model weights, autoregressive sampling, or incremental streaming codec math.

### MOSS-TTS Delay

A full SeedTTS EN evaluation was run on the same speed-selected paired rounds `[1, 3, 4, 6, 7]`, with 1088 samples per variant in every round. Quality metrics were not used to select the rounds. Raw results are shown as `Base / Current`:

| Round | Corpus WER | Corpus WER, excluding samples above 50% WER | Speaker Similarity | UTMOS |
|---:|---:|---:|---:|---:|
| 1 | 1.3983% / 1.6328% | 1.2739% / 1.2774% | 68.9948 / 68.8251 | 3.9413 / 3.9367 |
| 3 | 1.6997% / 1.4988% | 1.3120% / 1.3011% | 69.2043 / 68.8643 | 3.9420 / 3.9417 |
| 4 | 1.6579% / 1.6495% | 1.3773% / 1.4352% | 68.8058 / 69.0682 | 3.9346 / 3.9353 |
| 6 | 1.9174% / 1.4402% | 1.2232% / 1.2497% | 69.0203 / 68.9251 | 3.9402 / 3.9403 |
| 7 | 1.4151% / 1.5909% | 1.3571% / 1.3187% | 68.7677 / 68.8017 | 3.9478 / 3.9380 |

| Metric | Base mean | Current mean | Current - Base paired 95% CI |
|---|---:|---:|---:|
| Corpus WER | 1.6177% | 1.5624% | `[-0.416654, +0.306129]` pp |
| Corpus WER, excluding samples above 50% WER | 1.3087% | 1.3164% | `[-0.037765, +0.053140]` pp |
| Speaker Similarity | 68.9586 | 68.8969 | `[-0.342364, +0.219055]` |
| UTMOS | 3.9412 | 3.9384 | `[-0.008303, +0.002743]` |

All four paired confidence intervals include zero, indicating no statistically detectable quality change in this five-round evaluation.

### MOSS-TTS Local non-streaming

The shared decoder also changes the Local non-streaming path, so its full SeedTTS EN outputs were evaluated on the corresponding speed-selected paired rounds `[1, 2, 3, 4, 7]`, with 1088 samples per variant in every round. Quality metrics were not used to select the rounds. Raw results are shown as `Base / Current`:

| Round | Corpus WER | Corpus WER, excluding samples above 50% WER | Speaker Similarity | UTMOS |
|---:|---:|---:|---:|---:|
| 1 | 1.7249% / 2.0179% | 1.4690% / 1.5394% | 65.3937 / 65.2521 | 3.9627 / 3.9518 |
| 2 | 1.6830% / 2.4282% | 1.4189% / 1.5228% | 65.3076 / 65.5039 | 3.9614 / 3.9611 |
| 3 | 2.1519% / 2.3026% | 1.5344% / 1.5185% | 65.2754 / 65.5063 | 3.9533 / 3.9580 |
| 4 | 2.2105% / 2.0514% | 1.7086% / 1.6143% | 65.0906 / 65.0762 | 3.9497 / 3.9537 |
| 7 | 2.1519% / 2.1938% | 1.7001% / 1.6584% | 65.1965 / 65.3121 | 3.9609 / 3.9660 |

| Metric | Base mean | Current mean | Current - Base paired 95% CI |
|---|---:|---:|---:|
| Corpus WER | 1.9844% | 2.1988% | `[-0.207294, +0.635997]` pp |
| Corpus WER, excluding samples above 50% WER | 1.5662% | 1.5707% | `[-0.096699, +0.105656]` pp |
| Speaker Similarity | 65.2528 | 65.3301 | `[-0.114559, +0.269266]` |
| UTMOS | 3.9576 | 3.9581 | `[-0.007847, +0.008887]` |

All four paired confidence intervals include zero, indicating no statistically detectable Local quality change in this five-round evaluation.

### Codec numerical parity and tests

A production-shaped codec A/B microbenchmark used eight mixed segments with lengths `[33, 42, 56, 64, 75, 80, 92, 100]` and 60 interleaved repetitions on one NVIDIA H200. The retained implementation matched the source decoder's eight output lengths and produced bitwise-identical FP32 waveforms for every comparison.

Numerical tests cover exact inverse-delay reconstruction, quantizer equivalence, SDPA fallback, packed FlashAttention parity, local-causal window parity, exact RoPE on the actual non-contiguous Q/K layout, mixed-length packing, output lengths, segment ordering, and standalone fallback behavior.

### Vocoder hot-path microbenchmark

- Device: one NVIDIA H200.
- Input: eight mixed codec segments with lengths `[33, 42, 56, 64, 75, 80, 92, 100]`.
- Method: 60 interleaved A/B repetitions; each retained step preserves waveform lengths and bitwise FP32 output.

| Variant | Median wall time | Speedup vs baseline | Waveform |
|---|---:|---:|---:|
| Baseline | 50.663 ms | 1.000x | exact |
| Transfer and synchronization cleanup | 45.001 ms | 1.126x | exact |
| Cached quantizer | 43.440 ms | 1.166x | exact |
| Exact fused RoPE | 33.120 ms | 1.530x | exact |
| Full retained implementation | 31.601 ms | **1.603x** | exact |

The delay-pattern helper decreased from 241.45 us to 2.56 us for the measured input. The final profiler attributes the largest remaining GPU shares to FP32-to-BF16 autocast copies, packed FlashAttention, LayerNorm, layer scale, residual addition, and GELU; candidates that changed waveform numerics were rejected rather than included for additional speed.

### MOSS-TTS Delay end-to-end

- Device: one NVIDIA H200 143 GB.
- Model: `OpenMOSS-Team/MOSS-TTS-v1.5`.
- Workload: full SeedTTS EN, 1088/1088 requests, concurrency 16, non-streaming voice cloning, `references`, token count `auto`, one warmup request, and a fresh offline service for each variant.
- Integrity: every variant and round completed 1088/1088 requests with zero failures. Seven complete Base/Current candidate pairs were retained.

Raw candidate results are shown as `Base / Current`:

| Round | Selected | Throughput QPS | Mean latency (s) | P95 latency (s) | Mean RTF |
|---:|:---:|---:|---:|---:|---:|
| 1 | Yes | 7.531 / 8.457 | 2.116 / 1.884 | 2.563 / 2.338 | 0.4927 / 0.4355 |
| 2 | No | 7.578 / 8.500 | 2.103 / 1.875 | 2.559 / 2.321 | 0.4896 / 0.4335 |
| 3 | Yes | 7.584 / 8.558 | 2.101 / 1.862 | 2.557 / 2.314 | 0.4887 / 0.4304 |
| 4 | Yes | 7.623 / 8.516 | 2.090 / 1.872 | 2.537 / 2.317 | 0.4862 / 0.4349 |
| 5 | No | 7.525 / 8.555 | 2.117 / 1.863 | 2.560 / 2.317 | 0.4924 / 0.4306 |
| 6 | Yes | 7.551 / 8.566 | 2.111 / 1.860 | 2.579 / 2.310 | 0.4914 / 0.4300 |
| 7 | Yes | 7.577 / 8.503 | 2.104 / 1.874 | 2.564 / 2.342 | 0.4895 / 0.4332 |

Five-round subsets were enumerated in round order. Base and Current were screened independently on QPS, mean latency, P95 latency, and mean RTF using `abs(modified-Z) > 3.5` or the Tukey `3xIQR` extreme fence. `[1, 3, 4, 6, 7]` was the first eligible subset.

| Metric | Base mean | Current mean | Change | Current - Base paired 95% CI |
|---|---:|---:|---:|---:|
| Throughput QPS | 7.573200 | 8.520000 | **+12.50%** | `[+0.887411, +1.006189]` QPS |
| Mean latency | 2.104400 s | 1.870400 s | **-11.12%** | `[-0.249080, -0.218920]` s |
| P95 latency | 2.560000 s | 2.324200 s | **-9.21%** | `[-0.261479, -0.210121]` s |
| Mean RTF | 0.489700 | 0.432800 | **-11.62%** | `[-0.061463, -0.052337]` |

The paired confidence intervals are two-sided Student-t intervals over Current-minus-Base round differences with four degrees of freedom.

### MOSS-TTS Local non-streaming impact

The shared decoder also changes the Local non-streaming import path, so it was evaluated separately with the full SeedTTS EN workload on the same device class and concurrency. Seven complete Base/Current candidate pairs were retained; every candidate completed 1088/1088 requests with zero failures.

Raw candidate results are shown as `Base / Current`:

| Round | Selected | Throughput QPS | Mean latency (s) | P95 latency (s) | Mean RTF |
|---:|:---:|---:|---:|---:|---:|
| 1 | Yes | 9.913 / 10.700 | 1.606 / 1.487 | 2.147 / 2.030 | 0.3784 / 0.3504 |
| 2 | Yes | 9.902 / 10.663 | 1.608 / 1.492 | 2.175 / 2.017 | 0.3785 / 0.3514 |
| 3 | Yes | 9.953 / 10.551 | 1.599 / 1.510 | 2.169 / 2.055 | 0.3774 / 0.3558 |
| 4 | Yes | 9.967 / 10.795 | 1.597 / 1.475 | 2.135 / 1.995 | 0.3757 / 0.3477 |
| 5 | No | 9.932 / 10.739 | 1.603 / 1.483 | 2.148 / 2.064 | 0.3783 / 0.3498 |
| 6 | No | 10.023 / 10.712 | 1.588 / 1.487 | 2.127 / 1.987 | 0.3738 / 0.3504 |
| 7 | Yes | 9.946 / 10.812 | 1.600 / 1.472 | 2.153 / 1.982 | 0.3774 / 0.3471 |

Five-round subsets were enumerated in round order. Base and Current were screened independently on QPS, mean latency, P95 latency, and mean RTF using `abs(modified-Z) > 3.5` or the Tukey `3xIQR` extreme fence; accuracy metrics were not used for selection. `[1, 2, 3, 4, 7]` was the first eligible subset.

| Metric | Base mean | Current mean | Change | Current - Base paired 95% CI |
|---|---:|---:|---:|---:|
| Throughput QPS | 9.936200 | 10.704200 | **+7.73%** | `[+0.639991, +0.896009]` QPS |
| Mean latency | 1.602000 s | 1.487200 s | **-7.17%** | `[-0.133536, -0.096064]` s |
| P95 latency | 2.155800 s | 2.015800 s | **-6.49%** | `[-0.170979, -0.109021]` s |
| Mean RTF | 0.377480 | 0.350480 | **-7.15%** | `[-0.031026, -0.022974]` |

## Checklist

- [x] Format your code according with pre-commit.
- [x] Add unit tests.
- [x] Update documentation / docstrings / example tutorials as needed.
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the
`run-ci` label. Once labeled, every subsequent push re-triggers CI as
long as the label remains. Use `/tag-and-rerun-ci higgs` or
`/tag-and-rerun-ci moss` to select a TTS CI model, and
`/tag-and-rerun-ci fun-asr` or `/tag-and-rerun-ci qwen3-asr` to select an ASR
CI model. One selector from each family can be combined, for example
`/tag-and-rerun-ci moss fun-asr`. Draft PRs are skipped even if labeled.

### CODE REVIEW

File: sglang_omni/models/moss_tts_local/streaming_vocoder.py [hunk 1]
```diff
@@ -16,8 +16,8 @@
 
 import torch
```
Comment (luojiaxuan) [inline sglang_omni/models/moss_tts_local/streaming_vocoder.py:17]: Could we keep sglang_omni.models.moss_tts_local.vocoder_decoder as a small re-export shim? The move otherwise breaks downstream imports of MossTTSLocalVocoderDecoder.

File: sglang_omni/models/moss_tts/vocoder_decoder.py [hunk 2]
```diff
@@ -311,26 +447,58 @@ def __init__(
...
-            and self._can_run_packed_flash(x)
-        ):
-            return "flash_attention_2"
-        return self.source.resolve_attention_implementation(x, is_streaming=False)
-
-    def _can_run_packed_flash(self, x: torch.Tensor) -> bool:
-        if self._flash_attn_varlen is None:
+        preferred = getattr(self.source, "attention_implementation", None)
+        if preferred is None:
+            # The legacy tokenizer has no backend selector, but exposes the same
```
Comment (luojiaxuan) [inline sglang_omni/models/moss_tts/vocoder_decoder.py]: add note prefix plz

File: sglang_omni/models/moss_tts/vocoder_kernels.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        sin = tl.load(
+            cos_sin + position * head_dim + half_dim + pair,
+            mask=mask,
+        ).to(tl.float32)
+        qr = tl.load(q_base, mask=mask).to(tl.float32)
+        qi = tl.load(q_base + 1, mask=mask).to(tl.float32)
+        kr = tl.load(k_base, mask=mask).to(tl.float32)
+        ki = tl.load(k_base + 1, mask=mask).to(tl.float32)
+
+        # The source implementation materializes each FP32 multiply before the
```
Comment (luojiaxuan) [inline sglang_omni/models/moss_tts/vocoder_kernels.py]: ditto

**Review by luojiaxuan — APPROVED** [review]

File: sglang_omni/models/moss_tts/vocoder_decoder.py (hunk 2 above)
Comment (CloudRipple) [reply sglang_omni/models/moss_tts/vocoder_decoder.py]: Done in 14ca8f8f. Added the Note (Zhang Yiyang): prefix and clarified the legacy MOSS-Audio-Tokenizer compatibility behavior.

File: sglang_omni/models/moss_tts/vocoder_kernels.py (hunk 3 above)
Comment (CloudRipple) [reply sglang_omni/models/moss_tts/vocoder_kernels.py]: Done as well in 14ca8f8f. Added the same Note (Zhang Yiyang): prefix to the FP32 rounding comment.
